### PR TITLE
Add SanitizeBoundingBoxes transform

### PR DESCRIPTION
**Pull request by NicolasHug**

### CODE REVIEW

File: torchvision/prototype/transforms/_misc.py [hunk 1]
```diff
@@ -225,28 +226,83 @@ def _transform(
...
         super().__init__()
         self.min_size = min_size
+        self.labels = labels
+
+    def _find_label_default_heuristic(self, inputs):
+        # Tries to find a "label" key, otherwise tries for the first key that contains "label" - case insensitive
+        # Returns None if nothing is found
+        candidate_key = None
+        with suppress(StopIteration):
+            candidate_key = next(key for key in inputs.keys() if key.lower() == "label")
```
Comment (pmeier) [inline torchvision/prototype/transforms/_misc.py:246]: Can we make that `"labels"`? This is what we use in our references:

https://github.com/pytorch/vision/blob/6af6bf45cc36455e119cbb600503c7d0c8700ff4/references/detection/coco_utils.py#L88

File: torchvision/prototype/transforms/_misc.py [hunk 2]
```diff
@@ -225,28 +226,83 @@ def _transform(
...
+        if candidate_key is None:
+            with suppress(StopIteration):
+                candidate_key = next(key for key in inputs.keys() if "label" in key.lower())
+        labels = inputs.get(candidate_key)
+        return labels
+
+    def forward(self, *inputs: Any) -> Any:
+        inputs = inputs if len(inputs) > 1 else inputs[0]
+        if isinstance(labels, str) and not isinstance(inputs, dict):
```
Comment (pmeier) [inline torchvision/prototype/transforms/_misc.py]:
```suggestion
        if isinstance(self.labels, str) and not isinstance(inputs, dict):
```

Comment (pmeier) [reply torchvision/prototype/transforms/_misc.py]: I guess checking against dict is fine for now, but in general we should be checking against `collections.abc.Mapping`, right?

File: torchvision/prototype/transforms/_misc.py [hunk 3]
```diff
@@ -225,28 +226,83 @@ def _transform(
...
+            labels = self._find_label_default_heuristic(inputs)
+        elif callable(self.labels):
+            labels = self.labels(inputs)
+        elif isinstance(self.labels, str):
+            labels = inputs[self.labels]
+        else:
+            raise ValueError(
+                "labels parameter should either be a str, callable, or 'default'. "
+                f"Got {labels} of type {type(labels)}."
+            )
```
Comment (pmeier) [inline torchvision/prototype/transforms/_misc.py:271]: SHould we handle this in the constructor? Meaning, create a single callable from all input types so that this is just `labels = self.labels(inputs)`?

File: torchvision/prototype/transforms/_misc.py [hunk 4]
```diff
@@ -225,28 +226,83 @@ def _transform(
...
+        keep &= (boxes[:, 1] < image_h).all() & (boxes[:, 3] < image_h).all()
+        valid_indices = torch.where(keep)[0]
+
+        params = dict(valid_indices=valid_indices, labels=labels)
+        flat_outputs = [
+            # Even-though it may look like we're transforming all inputs, we don't:
+            # _transform() will only care about BoundingBoxes and the labels
+            self._transform(inpt, params)
+            for inpt in flat_inputs
+        ]
```
Comment (pmeier) [inline torchvision/prototype/transforms/_misc.py:299]: Not sure if we can do better without other changes, but this looks pretty weird. I mean, we have the bounding box and labels here. All we need to do is to put it at the right place in `flat_inputs` and we should be good to go without going the extra mile through `self._transform`.

**Review by pmeier — COMMENTED** [review]
Thanks Nicolas. Left a few high level comments.

File: torchvision/prototype/transforms/_misc.py (hunk 4 above)
Comment (NicolasHug) [reply torchvision/prototype/transforms/_misc.py:299]: > All we need to do is to put it at the right place in flat_inputs

yup... and I don't know how to do that easily :)
But if you can find a way to bypass `_transforms()`, I'm all ears

Comment (pmeier) [reply torchvision/prototype/transforms/_misc.py:299]: I mean for boxes, we can change `query_bounding_box` to whatever we like. Meaning, we could return the index from there. For the labels the story is different. We can't pass the `flat_inputs` because we rely on the dict keys. Meaning, users would need to return a spec similar to what `tree_flatten` produces, but that is bad UX. :shrug:

File: test/test_prototype_transforms.py [hunk 5]
```diff
@@ -2355,3 +2356,62 @@ def test_detection_preset(image_type, label_type, data_augmentation, to_tensor):
...
+
+    assert out["image"] is sample["image"]
+    assert out["whatever"] is sample["whatever"]
+    assert type(out["labels"]) is type(sample["labels"])
+
+    out["labels"] = torch.tensor(out["labels"])
+    assert out["boxes"].shape[0] == out["labels"].shape[0]
+
+    # This works because we conveniently set labels to arange(num_boxes)
+    assert out["labels"].tolist() == valid_indices
```
Comment (NicolasHug) [inline test/test_prototype_transforms.py:2411]: I guess I can also manually check that all valid_boxes are there, and that there's no invalid ones... LMK

File: torchvision/prototype/transforms/_misc.py [hunk 6]
```diff
@@ -225,28 +226,99 @@ def _transform(
...
-class RemoveSmallBoundingBoxes(Transform):
-    _transformed_types = (datapoints.BoundingBox, datapoints.Mask, datapoints.Label, datapoints.OneHotLabel)
+class SanitizeBoundingBoxes(Transform):
+    # This removes boxes and their corresponding labels:
+    # - small or degenerate bboxes based on min_size (this includes those where X2 <= X1 or Y2 <= Y1)
+    # - boxes with any coordinate outside the range of the image (negative, or > spatial_size)
+    _transformed_types = datapoints.BoundingBox
 
-    def __init__(self, min_size: float = 1.0) -> None:
+    def __init__(self, min_size: float = 1.0, labels="default") -> None:
```
Comment (pmeier) [inline torchvision/prototype/transforms/_misc.py:235]: "default" is pretty expressive, but maybe `None` is a good enough sentinel here? The other options are str and callable so that could be an option.

**Review by pmeier — COMMENTED** [review]
Looks good so far. We should also test the other label heuristics though.

One thing that came to mind is that only allowing a str key might be a little restrictive. Even for our builtin pipelines we have `(PIL.Image.Image, dict(labels=...))`. Maybe we can relax that and accept a sequence which items are used for a repeated index? Meaning, if someone passes `labels=(1, "labels")`, we do

```python
input = input[1]
input = input["labels"]
```

File: torchvision/prototype/transforms/_misc.py (hunk 6 above)
Comment (NicolasHug) [reply torchvision/prototype/transforms/_misc.py:235]: That reminds me I didn't go for `None` because we should allow `labels=None`, which means "I don't have labels and I don't care" - that should be allowed right?

**Comment by NicolasHug** [comment]
> Maybe we can relax that and accept a sequence which items are used for a repeated index?

I had in mind something similar where we would go as deep as needed in the inputs until we find a `"labels"` key (up to case-insensitive, etc.). This is similar to yours except that the "nesting" could be handled automatically. I'm happy to do that if we can do it easily, otherwise we always allow to pass callables so I'd say it's alright

File: torchvision/prototype/transforms/_misc.py [hunk 7]
```diff
@@ -225,28 +226,99 @@ def _transform(
...
+        return labels
+
+    def forward(self, *inputs: Any) -> Any:
+        inputs = inputs if len(inputs) > 1 else inputs[0]
+
+        if isinstance(self.labels, str) and not isinstance(inputs, dict):
+            raise ValueError(
+                f"If labels is a str or 'default' (got {self.labels}), then the input to forward() must be a dict. "
+                f"Got {type(inputs)} instead"
+            )
```
[the remark this answers is not in the document]
Comment (NicolasHug) [reply torchvision/prototype/transforms/_misc.py:281]: Maybe I don't understand what you have in mind, but we have to do this check here (in fowrad()) anyway because it must have the `dict()`. To this end, I'd rather having it explicitly here than "monkey-patch" that check into both `_find_label_default_heuristic` and the `lambda`?

File: torchvision/prototype/transforms/_misc.py (hunk 6 above)
Comment (pmeier) [reply torchvision/prototype/transforms/_misc.py:235]: Fair enough.

File: torchvision/prototype/transforms/_misc.py [hunk 8]
```diff
@@ -225,28 +226,99 @@ def _transform(
...
+
+        boxes = F.convert_format_bounding_box(
+            boxes,
             new_format=datapoints.BoundingBoxFormat.XYXY,
         )
-        valid_indices = remove_small_boxes(bounding_box, min_size=self.min_size)
-
-        return dict(valid_indices=valid_indices)
+        ws, hs = boxes[:, 2] - boxes[:, 0], boxes[:, 3] - boxes[:, 1]
```
[the remark this answers is not in the document]
Comment (NicolasHug) [reply torchvision/prototype/transforms/_misc.py:301]: Thinking about this more I don't think it is possible to support multiple batches. Say we remove 3 boxes in the first batch entry and 2 boxes in the second batch entry: we end up with a batch with entries of different size - that's not supported.

I'll still add support for batch with 1 element, just in case (and error otherwise)

File: torchvision/prototype/transforms/_misc.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (pmeier) [reply torchvision/prototype/transforms/_misc.py]: Unless I miss something, this check is only here for the case of user supplied string keys, right? Can't we replace the lambda in the constructor with

```python
def _find_label_str_heuristic(inputs):
    if not isinstance(inputs, dict):
        raise ValueError(
            f"If labels is a str or 'default' (got {self.labels}), then the input to forward() must be a dict. "
            f"Got {type(inputs)} instead"
        )

     return inputs[self.labels]
```

[the remark this answers is not in the document]
Comment (NicolasHug) [reply torchvision/prototype/transforms/_misc.py]: > this check is only here for the case of user supplied string keys

The check is valid for `"default"` as well

File: torchvision/prototype/transforms/_misc.py [hunk 9]
```diff
@@ -225,28 +227,111 @@ def _transform(
...
+            raise ValueError(f"The labels in the input to forward() must be a tensor, got {type(labels)} instead.")
+
+        flat_inputs, spec = tree_flatten(inputs)
+        # TODO: this enforces one single BoundingBox entry.
+        # Assuming this transform needs to be called at the end of *any* pipeline that has bboxes...
+        # should we just enforce it for all transforms?? What are the benefits of *not* enforcing this?
+        boxes = query_bounding_box(flat_inputs)
+
+        if boxes.ndim > 3 or (boxes.ndim == 3 and boxes.shape[0] != 1):
+            raise ValueError(f"boxes must be of shape (num_boxes, 4) or (1, num_boxes, 4), got {boxes.shape}")
```
Comment (pmeier) [inline torchvision/prototype/transforms/_misc.py:298]: So far we don't have a use case for stacked boxes. Since we can't support arbitrary batches here anyway, is there a good reason to allow `(1, N, 4)`? If not, let's just enforce `(N, 4)`.

File: torchvision/prototype/transforms/_misc.py [hunk 10]
```diff
@@ -225,28 +227,111 @@ def _transform(
...
+            )
+
+        boxes = F.convert_format_bounding_box(
+            boxes,
             new_format=datapoints.BoundingBoxFormat.XYXY,
         )
-        valid_indices = remove_small_boxes(bounding_box, min_size=self.min_size)
-
-        return dict(valid_indices=valid_indices)
+        ws, hs = boxes[..., 2] - boxes[..., 0], boxes[..., 3] - boxes[..., 1]
```
Comment (pmeier) [inline torchvision/prototype/transforms/_misc.py:309]: Nit: if we enforce 2 dims only, we could go back to 

```suggestion
        ws, hs = boxes[:, 2] - boxes[:, 0], boxes[:, 3] - boxes[:, 1]
```

to make that more clear.

File: test/test_prototype_transforms.py [hunk 11]
```diff
@@ -2355,3 +2356,124 @@ def test_detection_preset(image_type, label_type, data_augmentation, to_tensor):
...
+        "whatever": torch.rand(10),
+        "None": None,
+    }
+
+    out = transforms.SanitizeBoundingBoxes(min_size=min_size, labels_getter=labels_getter)(sample)
+
+    assert out["image"] is sample["image"]
+    assert out["whatever"] is sample["whatever"]
+
+    if labels_getter is None or (callable(labels_getter) and labels_getter({"labels": "blah"}) is None):
```
Comment (pmeier) [inline test/test_prototype_transforms.py:2418]: I see that returning `None` from callable labels getter is intentional. Why though? If someone explicitly sets this, why would they want to (silently) not transform labels? Imagine, someone just forgets a `return` statement (which happens regularly but not often to me).

Comment (NicolasHug) [reply test/test_prototype_transforms.py:2418]: what if the label is from the image and not the boxes?

File: torchvision/prototype/transforms/_misc.py [hunk 12]
```diff
@@ -225,28 +227,111 @@ def _transform(
...
+            with suppress(StopIteration):
+                candidate_key = next(key for key in inputs.keys() if "label" in key.lower())
+        labels = inputs.get(candidate_key)
+
+        if labels is None:
+            raise ValueError(
+                "Could not infer where the labels are in the sample. Try passing a callable as the labels_getter parameter?"
+                "If there are no samples and it is by design, pass labels_getter=None."
+            )
+        return labels
```
[the remark this answers is not in the document]
Comment (NicolasHug) [reply torchvision/prototype/transforms/_misc.py:276]: That doesn't work, `labels` is always None in your snippet. Did you mean `candidate_key`?

File: torchvision/prototype/transforms/_misc.py (hunk 9 above)
Comment (NicolasHug) [reply torchvision/prototype/transforms/_misc.py:298]: I added it after your suggestion to support batches. We can remove it, but let's merge it first to unblock and come back to this if needed

File: test/test_prototype_transforms.py (hunk 11 above)
Comment (pmeier) [reply test/test_prototype_transforms.py:2418]: They should pass `labels=None`? Unless I miss something, this only applies to user defined callables, since the default heuristic errors in case it doesn't find anything. So why would a user use `labels=lambda: None` if that has the same behavior as `labels=None`? The only reason I see is if the callable can conditionally return `None` or the labels. But I fail to see a use case for that.

File: torchvision/prototype/transforms/_misc.py (hunk 9 above)
Comment (NicolasHug) [reply torchvision/prototype/transforms/_misc.py:298]: Ah, F it, I'll remove it. Supporting a batch dim made the code a lot worse

File: test/test_prototype_transforms.py (hunk 11 above)
Comment (NicolasHug) [reply test/test_prototype_transforms.py:2418]: what if they have some samples with labels and some without?

File: torchvision/prototype/transforms/_misc.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (NicolasHug) [reply torchvision/prototype/transforms/_misc.py:276]: done

File: test/test_prototype_transforms.py (hunk 11 above)
Comment (NicolasHug) [reply test/test_prototype_transforms.py:2412]: imagine the samples are

```
labels_getter inputs = inputs[2]
# train
(img, bboxes, labels)
# test? Or some other weird stuff??
(img, bboxes, None)
```
?

Comment (pmeier) [reply test/test_prototype_transforms.py:2412]: WE discussed offline and I just understood the mechanism wrong. Everything good as is.

**Comment by NicolasHug** [comment]
Thanks a lot for the reviews and for the help with `myp`
